# Contents List

## 01. Importing Libraries
## 02. Paths
## 03. Importing Data
## Task 4.6 B

# 01. Importing libraries

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 2. Paths

In [3]:
#Creating a path varaible - Shortcut for Importing Data Files
path = r'/Users/aya/Desktop/DataAnalytics/Data-Immersion/Achievement-4/Instacart-Basket-Analysis'

In [4]:
path

'/Users/aya/Desktop/DataAnalytics/Data-Immersion/Achievement-4/Instacart-Basket-Analysis'

# 03. Importing Data

In [9]:
#Import orders_products_combined Dataset
orders_products_combined = pd.read_pickle(os.path.join(path, '02-Data', 'Prepared-Data', 'orders_products_combined.pkl'))

In [10]:
#Checking the dataframe -orders_products_combined
orders_products_combined.shape

(32434489, 10)

In [11]:
#Import orders_products_combined Dataset
df_prods = pd.read_csv(os.path.join(path, '02-Data', 'Prepared-Data', 'products_checked.csv'), index_col = False)

In [12]:
#Checking the dataframe -df_prods
df_prods.shape

(49672, 6)

### Observation 3. & 4.
##### Data imported and checked
##### The dataframe orders_products_combined contains 32434489 rows and 10 columns

In [18]:
#viewing the df_prods dataframe
df_prods

Unnamed: 0  product_id  \
0               0           1   
1               1           2   
2               2           3   
3               3           4   
4               4           5   
...           ...         ...   
49667       49688       49684   
49668       49689       49685   
49669       49690       49686   
49670       49691       49687   
49671       49692       49688   

                                            product_name  aisle_id  \
0                             Chocolate Sandwich Cookies        61   
1                                       All-Seasons Salt       104   
2                   Robust Golden Unsweetened Oolong Tea        94   
3      Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4                              Green Chile Anytime Sauce         5   
...                                                  ...       ...   
49667          Vodka, Triple Distilled, Twist of Vanilla       124   
49668                 En Croute Roast Hazelnut Cranberry        42   
49669                                   Artisan Baguette       112   
49670         Smartblend Healthy Metabolism Dry Cat Food        41   
49671                             Fresh Foaming Cleanser        73   

       department_id  prices  
0                 19     5.8  
1                 13     9.3  
2                  7     4.5  
3                  1    10.5  
4                 13     4.3  
...              ...     ...  
49667              5     5.3  
49668              1     3.1  
49669              3     7.8  
49670              8     4.7  
49671             11    13.5  

[49672 rows x 6 columns]

In [20]:
#Dropping the column "Unnamed: 0" from the df_prods dataframe and assinged it to the "df_prods" dataframe
df_prods = df_prods.drop('Unnamed: 0', axis=1)

In [22]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49667       49684          Vodka, Triple Distilled, Twist of Vanilla   
49668       49685                 En Croute Roast Hazelnut Cranberry   
49669       49686                                   Artisan Baguette   
49670       49687         Smartblend Healthy Metabolism Dry Cat Food   
49671       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49667       124              5     5.3  
49668        42              1     3.1  
49669       112              3     7.8  
49670        41              8     4.7  
49671        73             11    13.5  

[49672 rows x 5 columns]

In [24]:
#Checking the orders_products_combined dataframe for the common columns with the df_prods dataframe
orders_products_combined

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2539329        1             1                   2   
2          2539329        1             1                   2   
3          2539329        1             1                   2   
4          2539329        1             1                   2   
...            ...      ...           ...                 ...   
32434484   2977660   206209            13                   1   
32434485   2977660   206209            13                   1   
32434486   2977660   206209            13                   1   
32434487   2977660   206209            13                   1   
32434488   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         8                     NaN       14084   
2                         8                     NaN       12427   
3                         8                     NaN       26088   
4                         8                     NaN       26405   
...                     ...                     ...         ...   
32434484                 12                     7.0       14197   
32434485                 12                     7.0       38730   
32434486                 12                     7.0       31477   
32434487                 12                     7.0        6567   
32434488                 12                     7.0       22920   

          add_to_cart_order  reordered _merge  
0                         1          0   both  
1                         2          0   both  
2                         3          0   both  
3                         4          0   both  
4                         5          0   both  
...                     ...        ...    ...  
32434484                  5          1   both  
32434485                  6          0   both  
32434486                  7          0   both  
32434487                  8          0   both  
32434488                  9          0   both  

[32434489 rows x 10 columns]

In [26]:
#Create a new dataframe to merge the orders_products_combined dataframe and df_prods dataframe. 
ords_prods_merge = df_prods.merge(orders_products_combined, on='product_id', how ='inner')

In [27]:
#Checking the new dataframe
ords_prods_merge.shape

(32404859, 14)

### Observation 5.

##### After viewing the df_prods dataframe, I noticed an extra "Unnamed: 0" column that was not needed. 
##### Removed this column, thus bringing the total number of rows in the df_prods dataframe to 49672 rows × 5 columns. 


##### Searched for a common column to merge the df_prods and orders_products_combined dataframes.
##### Found that both dataframes had the column 'product_id'.


##### The df_prods dataframe contains 49672 rows and 5 columns.
##### The orders_products_combined dataframe contains 32434489 rows x 10 columns.


##### Deciding on how to combine the data
##### Concatenation is not a good choice because both data sets are not of the same length.

##### Full outer join might not be the most suitable approach because the data frames are of different sizes. 
##### The left or right join may be not the best option because one of the dataframes will be joined only if there is a matche. This approach may result in the loss of some data from one of the dataframes, while other data maybe missing. 


##### Inner join is the choice that make sense. 
##### That's because we need the product information mapped to the orders_products_combined dataframe. 
##### That way, it helps us to answer many business questions about customers and their shopping behaviors. 
##### This approach results in a merge where only the rows that have a match in both dataframes will be included in the final merged dataframe. All other rows will be discarded.

##### Create a new dataframe, ords_prods_merge, to merge the orders_products_combined dataframe and df_prods dataframe together, using inner join.

In [30]:
#Checking the merge
ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

### Observation 6.
##### The new dataframe, ords_prods_merge, contains 32404859 rows X 14 columns.
##### The total merged rows is 32404859.

In [120]:
# Export the df_merged_large dataframe to pkl formate
ords_prods_merge.to_pickle(os.path.join(path, '02-Data','Prepared-Data', 'ords_prods_merge.pkl'))

### Observation 7.
##### Export the merged data, in pickle format as “ords_prods_merge.pkl” 